rust/main/chains/hyperlane-ethereum/src/rpc_clients/fallback/tests.rs

```diff
@@ -0,0 +1,216 @@
+    let providers = vec![
+        EthereumProviderMock::new(Some(Duration::from_millis(10)), true),
+        EthereumProviderMock::new(None, false),
+        EthereumProviderMock::default(),
+    ];
+    let fallback_provider = fallback_provider_builder.add_providers(providers).build();
+    let ethereum_fallback_provider = EthereumFallbackProvider::new(fallback_provider);
+    ethereum_fallback_provider
+        .multicast::<_, u64>(BLOCK_NUMBER_RPC, ())
+        .await
+        .unwrap();
+    let provider_call_count: Vec<_> =
+        ProviderMock::get_call_counts(&ethereum_fallback_provider).await;
+    assert_eq!(provider_call_count, vec![0, 1, 1]);
+}
+
+#[tokio::test]
+async fn test_multicast_none_provider_succeeds() {
+    let fallback_provider_builder = FallbackProviderBuilder::default();
+    let providers = vec![
+        EthereumProviderMock::new(None, false),
+        EthereumProviderMock::new(None, false),
+        EthereumProviderMock::new(None, false),
+    ];
+    let fallback_provider = fallback_provider_builder.add_providers(providers).build();
+    let ethereum_fallback_provider = EthereumFallbackProvider::new(fallback_provider);
+    let _ = ethereum_fallback_provider
+        .multicast::<_, u64>(BLOCK_NUMBER_RPC, ())
+        .await;
+    let provider_call_count: Vec<_> =
+        ProviderMock::get_call_counts(&ethereum_fallback_provider).await;
+    assert_eq!(provider_call_count, vec![4, 4, 4]);
+}
+
+#[tokio::test]
+async fn test_fallback_first_provider_is_attempted() {
+    let fallback_provider_builder = FallbackProviderBuilder::default();
+    let providers = vec![
+        EthereumProviderMock::default(),
+        EthereumProviderMock::default(),
+        EthereumProviderMock::default(),
+    ];
+    let fallback_provider = fallback_provider_builder.add_providers(providers).build();
+    let ethereum_fallback_provider = EthereumFallbackProvider::new(fallback_provider);
+    ethereum_fallback_provider.fallback_test_call().await;
+    let provider_call_count: Vec<_> =
+        ProviderMock::get_call_counts(&ethereum_fallback_provider).await;
+    assert_eq!(provider_call_count, vec![1, 0, 0]);
+}
+
+#[tokio::test]
+async fn test_fallback_one_stalled_provider() {
+    let fallback_provider_builder = FallbackProviderBuilder::default();
+    let providers = vec![
+        EthereumProviderMock::new(Some(Duration::from_millis(10)), true),
+        EthereumProviderMock::default(),
+        EthereumProviderMock::default(),
+    ];
+    let fallback_provider = fallback_provider_builder
+        .add_providers(providers)
+        .with_max_block_time(Duration::from_secs(0))
+        .build();
+    let ethereum_fallback_provider = EthereumFallbackProvider::new(fallback_provider);
+    ethereum_fallback_provider.fallback_test_call().await;
+    let provider_call_count: Vec<_> =
+        ProviderMock::get_call_counts(&ethereum_fallback_provider).await;
+    assert_eq!(provider_call_count, vec![0, 0, 2]);
```

Comment:
I know this isn't new but why is this `vec![0, 0, 2]` instead of `vec![0, 1, 0]`?

Reply:
Since the first provider gets deprioritised, it is moved from 0th to 2nd position, so, the last entry in `provider_call_count` corresponds to the first provider. 

The value `2` instead of `1` is explained by the implementation of `handle_stalled_provider` where we query block number to identify if the provider stalled. So, it just happed that the kind of query we implemented the mock providers with is the same as the kind of query we identify stalled providers.

Reply:
I'll add a comment into the test case.